Chore: (Docs) Storybook 6.5 testing updates (#17967)
storybookjs/storybook: docs/writing-tests/accessibility-testing.md

Accessibility tests audit the rendered DOM against a set of heuristics based on [WCAG](https://www.w3.org/WAI/standards-guidelines/wcag/) rules and other industry-accepted best practices. They act as the first line of QA to catch blatant accessibility violations.

![Accessibility testing](./accessibility-testing-storybook.gif)

## Accessibility testing with a11y addon

Storybook provides an official [a11y addon](https://storybook.js.org/addons/@storybook/addon-a11y). Powered by Deque's [axe-core](https://github.com/dequelabs/axe-core), which automatically catches up to [57% of WCAG issues](https://www.deque.com/blog/automated-testing-study-identifies-57-percent-of-digital-accessibility-issues/).

### Set up the a11y addon

> If you want to run accessibility tests using the [addon](https://storybook.js.org/addons/@storybook/addon-a11y/), you'll need additional steps to set it up. Documented below is our recommendation.

Reviewer: > run accessibility tests

This phrase makes me think of automated testing, which is covered explicitly (and, importantly, _differently_) later. Maybe:

> check accessibility for your stories

Reviewer: > Documented below is our recommendation.

It's not so much a recommendation as "the way to do it", right? The way this section shares a lot of phrasing with the 'Automate accessibility tests with test runner' section made me wonder if there was a copy error for a moment.

Reviewer: So, we both of those comments in mind:

```suggestion
If you want to check accessibility for your stories using the [addon](https://storybook.js.org/addons/@storybook/addon-a11y/), you'll need to install it and add it to your Storybook.
```

Run the following command to install the addon.

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-a11y-install.yarn.js.mdx',
    'common/storybook-a11y-install.npm.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

> Update your Storybook configuration (in `.storybook/main.js|ts`) to include the accessibility addon.

Reviewer: Is it that common to have a `main.ts`? Doesn't seem worth the complication here.

Author: @kylegach yes, yes it is this was left here intentionally to avoid unwanted issues/questions about TS support as we had in the past, to which I responded and updated the documentation accordingly. The rest of the documentation will "fall in line" (pardon the bad pun) as well shortly

### Configure

> Out of the box, Storybook's accessibility addon includes a set of accessibility rules that cover most issues. You can also fine-tune the [addon configuration](https://github.com/storybookjs/storybook/tree/next/addons/a11y#parameters) or override [Axe's ruleset](https://github.com/storybookjs/storybook/tree/next/addons/a11y#handling-failing-rules) to best suit your needs.

Reviewer: I think, because this is configured via `parameters`, it will _not_ work when using axe via the test-runner. Do we need to call that out? I could definitely see people expect it to "just work".

And, outside of this docs PR, is there anything we can do to bridge the gap? /cc @yannbf

Reviewer 2: Yes. Discussed with @yannbf here https://github.com/storybookjs/test-runner/issues/97

Author: Thanks for the context @shilman! Appreciate it! I will follow up with both of you on this.

<!-- prettier-ignore-end -->

> ## Automate accessibility tests with test runner

Reviewer: I think this section needs to mention that automated a11y testing is just one, incomplete tool for validating your components. See: https://www.smashingmagazine.com/2018/09/importance-manual-accessibility-testing/

Reviewer: Of all my comments, this one would probably require the most thought. Please let me know if I can help.

Reviewer 3: How's this:

> The most accurate way to check accessibility is manually on real devices. However, you can use automated tools to catch common accessibility issues. For example, Axe, on average, catches [57% of WCAG issues automatically](https://www.deque.com/blog/automated-testing-study-identifies-57-percent-of-digital-accessibility-issues/).
>
> These tools work by auditing the rendered DOM against heuristics based on [WCAG](https://www.w3.org/WAI/standards-guidelines/wcag/) rules and other industry-accepted best practices.

Author: Updated based on the comment. Let me know if both of you are ok with this. So that we can tick this item off.

### Setup

To enable accessibility testing with the test runner, you will need to take additional steps to set it up properly. Detailed below is our recommendation to configure and execute them.

Run the following command to install the required dependencies.

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-test-runner-axe-playwright.yarn.js.mdx',
    'common/storybook-test-runner-axe-playwright.npm.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

Add a new [configuration file](./test-runner.md#test-hook-api-experimental) inside your Storybook directory with the following inside:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-test-runner-a11y-config.js.mdx',
    'common/storybook-test-runner-a11y-config.ts.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

<div class="aside">

💡 `preRender` and `postRender` are convenient hooks that allow you to extend the test runner's default configuration. They are **experimental** and subject to changes. Read more about them [here](./test-runner.md#test-hook-api-experimental).

</div>

> By default, Axe assumes that you're testing a page and checks whether you've specified an `<h1>` and `<main>`. However, most of your stories are for components and not pages. That's why we use the `context.title.includes('Page')` check to [enable/disable](https://github.com/abhinaba-ghosh/axe-playwright#context-optional) Axe's page-level rules.

Reviewer: ```suggestion
By default, Axe assumes that you're testing a page and checks whether you've specified an `<h1>` and `<main>`. However, not all of your stories are for page-level components. That's why we use the `context.title.includes('Page')` check to [enable/disable](https://github.com/abhinaba-ghosh/axe-playwright#context-optional) Axe's page-level rules.
```

> When you execute the test runner (i.e., `yarn test-storybook`), it will run the accessibility audit and any [interaction tests](./interaction-testing.md) you might have configured for each component story.

Reviewer: > i.e., `yarn test-storybook`

My understanding is that "i.e.", "e.g.", and other Latin abbreviations are not well understood by international audiences.

Author: This was addressed in the latest commits. Marking it as solved.

![Accessibility testing with the test runner](./test-runner-a11y-optimized.png)

---

#### What’s the difference between browser-based and linter-based accessibility tests?

> Browser-based accessibility tests, like found in Storybook, evaluate the rendered DOM because that gives you the highest accuracy. Auditing code that hasn't been compiled yet is one step removed from the real thing, so you won't catch everything the user might experience.

Reviewer: Appreciate this!

Author: No problem :)

These tools work by auditing the rendered DOM against heuristics based on [WCAG](https://www.w3.org/WAI/standards-guidelines/wcag/) rules and other industry-accepted best practices. You can then integrate these tools into your test automation pipeline using the Storybook [test runner](./test-runner.md#test-hook-api-experimental) and [axe-playwright](https://github.com/abhinaba-ghosh/axe-playwright).

### Setup

Accessibility testing with Storybook shortens the feedback loop which means you fix issues faster. Reuse stories in a Jest test and run an accessibility audit on them using the [jest-axe integration](https://github.com/nickcolley/jest-axe). That also unlocks the ability to integrate accessibility tests into your functional testing pipeline.
To enable accessibility testing with the test runner, you will need to take additional steps to set it up properly. Detailed below is our recommendation to configure and execute them.

For example, include the following test file to run an accessibility test on a story:
Run the following command to install the required dependencies.

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[

>     'react/accessibility-testing-with-jest-axe.js.mdx',

Reviewer 4: Kinda sucks that this is removed (lots of people might opt not to use the test runner yet and keep using testing-react), but I totally get why we would remove it from the docs.

Author: I hear you. I hear you, but look at it this way @yannbf we pique people's interest with it and help us improve it.

Add a new [configuration file](./test-runner.md#test-hook-api-experimental) inside your Storybook directory with the following inside:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-test-runner-a11y-config.js.mdx',
    'common/storybook-test-runner-a11y-config.ts.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

<div class="aside">

💡 `preRender` and `postRender` are convenient hooks that allow you to extend the test runner's default configuration. They are **experimental** and subject to changes. Read more about them [here](./test-runner.md#test-hook-api-experimental).

</div>

When you execute the test runner (for example, with `yarn test-storybook`), it will run the accessibility audit and any [interaction tests](./interaction-testing.md) you might have configured for each component story.

It starts checking for issues by traversing the DOM tree starting from the story's root element and generates a detailed report based on the issues it encountered.

![Accessibility testing with the test runner](./test-runner-a11y-optimized.png)

---

#### What’s the difference between browser-based and linter-based accessibility tests?

> Browser-based accessibility tests, like found in Storybook, evaluate the rendered DOM because that gives you the highest accuracy. Auditing code that hasn't been compiled yet is one step removed from the real thing, so you won't catch everything the user might experience.

Reviewer 2: ```suggestion
Browser-based accessibility tests, like those found in Storybook, evaluate the rendered DOM because that gives you the highest accuracy. Auditing code that hasn't been compiled yet is one step removed from the real thing, so you won't catch everything the user might experience.
```
